src/fb.c

```diff
@@ -1,36 +1,39 @@
 /* SPDX-License-Identifier: MIT */
 
+#include <assert.h>
+
 #include "fb.h"
+#include "string.h"
 #include "utils.h"
 #include "xnuboot.h"
 
-u32 *fb;
-int fb_s, fb_w, fb_h, fb_d;
-int scale = 1;
+#define CONSOLE_MAX_ROWS 60
+#define CONSOLE_MAX_COLS 80
+
+fb_t fb;
+
+typedef struct {
+    u32 row;
+    u32 col;
+} console_pos_t;
 
-static u32 *logo;
-static int logo_w, logo_h;
+static struct {
+    u32 *ptr;
+    int w;
+    int h;
+} logo;
+
+static struct {
+    u32 initialized;
+    console_pos_t cursor;
+    u32 row_offset;
+    u8 text[CONSOLE_MAX_ROWS][CONSOLE_MAX_COLS];
+} console = {.initialized = 0, .cursor = {.row = 0, .col = 0}, .row_offset = 0, .text = {{0}}};
 
 extern char _binary_build_bootlogo_128_bin_start[0];
 extern char _binary_build_bootlogo_256_bin_start[0];
 
-void fb_init(void)
-{
-    fb = (void *)cur_boot_args.video.base;
-    fb_s = cur_boot_args.video.stride / 4;
-    fb_w = cur_boot_args.video.width;
-    fb_h = cur_boot_args.video.height;
-    fb_d = cur_boot_args.video.depth & 0xff;
-    printf("fb init: %dx%d (%d) [s=%d] @%p\n", fb_w, fb_h, fb_d, fb_s, fb);
-
-    if (cur_boot_args.video.depth & 0x10000) {
-        logo = (void *)_binary_build_bootlogo_256_bin_start;
-        logo_w = logo_h = 256;
-    } else {
-        logo = (void *)_binary_build_bootlogo_128_bin_start;
-        logo_w = logo_h = 128;
-    }
-}
+extern u8 _binary_font_b_start[0];
```

Comment:
If you change the Makefile like I suggest you will need to rename this to `_binary_font_bin_start`